keps/sig-storage/5381-mutable-pv-affinity/README.md

```diff
@@ -0,0 +1,997 @@
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+`PersistentVolume.spec.nodeAffinity.required` should works like
+`Pod.spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution`.
+It is never re-evaluated when the pod is already running.
+It is storage provider's responsibility to ensure that the running workload is not interrupted.
+
+**Possible race condition**
+
+There is a race condition between volume modification and pod scheduling:
+1. User updates PVC VolumeAttributeClassName.
+2. external-resizer initiate ControllerModifyVolume.
+3. User creates a new Pod and scheduler schedules it with the old affinity.
+4. external-resizer sets a new affinity to the PV.
+5. KCM/external-attacher attaches the volume to the node, and find the affinity mismatch.
+
+If this happens, the pod will be stuck in a `ContainerCreating` state.
+User will have to manually delete the pod, or using Kubernetes [descheduler](https://github.com/kubernetes-sigs/descheduler) or similar.
+
+We propose to added an `in_progress` field in `ControllerModifyVolumeResponse` to allow more timely affinity update,
+in order to reduce the time window of the race condition.
+If the volume is not attachable during the modification, SPs can also return `in_progress`
+with an impossible topology requirement, so that new pods will stay pending.
+```json
+[{"provider.com/volume-status": "modifying"}]
+```
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We will extend CSI specification to add:
+```protobuf
+// Specifies a capability of the controller service.
+message ControllerServiceCapability {
+  message RPC {
+    enum Type {
+      ...
+      // TODO
+      MODIFY_VOLUME_TOPOLOGY = 16 [(alpha_enum_value) = true];
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+
+message ControllerModifyVolumeRequest {
+  option (alpha_message) = true;
+  ...
+
+  // Indicates whether the CO allows the SP to update the topology
+  // as a part of the modification.
+  bool allow_topology_updates = 4;
```

Comment:
We also need a new parameters in ```mutable_parameters``` right? Let's include an example demonstrating the usage of ControllerModifyVolumeRequest and ControllerModifyVolumeResponse with allow_topology_updates.